### ISSUE TITLE
.Skip() / .Take() generate problematic OFFSET inside subquery (EXISTS)

### ISSUE THREAD

**Issue by wc-matteo** [dotnet/efcore#6140]
### Steps to reproduce

https://github.com/aspnet/EntityFramework/issues/6140#issuecomment-234718460
### The issue

EF is generating this kind of query (taken from https://github.com/aspnet/EntityFramework/issues/5694#issuecomment-225307646) in my project and it's an issue (the query returns nothing). I'm including (with `.Include()`) a bunch of entities and paginating the results with `.Skip()` and `.Take()`.

`WHERE [c1].[CompanyId] = [c0].[CompanyId]` filters the join of the outer table (`[c1]`) with the one inside `EXISTS` (`[c0]`), but applying an `OFFSET` here is a mistake! It's should be put in the outer query, if at all (in my case, the outer query is for a secondary included entity; and it should certainly not have an offset!). The result of `SELECT 1` in `EXISTS` is gonna be on the first row, any `OFFSET` greater than 0 will make the `EXISTS` subquery return false.

``` sql
SELECT CASE WHEN EXISTS (SELECT * ORDER BY @@ROWCOUNT OFFSET 0 ROWS) THEN 1 ELSE 0 END
```

> 1

vs

``` sql
SELECT CASE WHEN EXISTS (SELECT * ORDER BY @@ROWCOUNT OFFSET 1 ROWS) THEN 1 ELSE 0 END
```

> 0

I suppose there isn't a hint or something to tell the framework where it should apply `.Skip()` / `.Take()`?
### Further technical details

EF Core version: 1.0.0
Operating system: Windows 10
Visual Studio version: n/a

**Comment by divega** [comment]
@wc-matteo could you include a full repro or at least the original LINQ query in the bug?

**Comment by wc-matteo** [comment]
### Model

##### Blogs

```cs
public class Blogs
{
    public List<Posts> Posts {get; set;}

    public int BlogsId { get; set; }
    public string Name { get; set; }
}
```

##### Posts

```cs
public class Posts
{
    public int PostsId { get; set; }
    public int BlogsId { get; set; }
    public string Title { get; set; }
}
```

### Controller

```cs
[Route("")]
public class Controller
{
    private readonly Context _dbContext;

    public Controller(Context dbContext)
    {
        _dbContext = dbContext;
    }

    private IEnumerable<string> FindTitles(Blogs blogs)
    {
        return blogs.Posts.Select(p => p.Title);
    }

    public object Get()
    {
        return _dbContext.Blogs
            .Include(b => b.Posts)
            .Select(b => new { BlogName = b.Name, PostTitles = FindTitles(b) })
            .Skip(1)
            .Take(1);
    }
}
```

### SQL Data

```sql
INSERT [dbo].[Blogs] ([BlogsId], [Name]) VALUES (1, N'Foo')
INSERT [dbo].[Blogs] ([BlogsId], [Name]) VALUES (2, N'Bar')
INSERT [dbo].[Blogs] ([BlogsId], [Name]) VALUES (3, N'Baz')

INSERT [dbo].[Posts] ([PostsId], [BlogsId], [Title]) VALUES (1, 1, N'Lorem')
INSERT [dbo].[Posts] ([PostsId], [BlogsId], [Title]) VALUES (2, 1, N'Ipsum')
INSERT [dbo].[Posts] ([PostsId], [BlogsId], [Title]) VALUES (3, 2, N'Dolor')
INSERT [dbo].[Posts] ([PostsId], [BlogsId], [Title]) VALUES (4, 3, N'Sit')
INSERT [dbo].[Posts] ([PostsId], [BlogsId], [Title]) VALUES (5, 3, N'Amet')
```

### Result

```json
[{"blogName":"Bar","postTitles":[]}]
```
### EF Generated SQL

```sql
SELECT [b].[BlogsId], [b].[Name]
FROM [Blogs] AS [b]
ORDER BY [b].[BlogsId]
OFFSET @__p_0 ROWS FETCH NEXT @__p_1 ROWS ONLY

SELECT [p].[PostsId], [p].[BlogsId], [p].[Title]
FROM [Posts] AS [p]
WHERE EXISTS (
    SELECT 1
    FROM [Blogs] AS [b]
    WHERE [p].[BlogsId] = [b].[BlogsId]
    ORDER BY @@ROWCOUNT
    OFFSET @__p_0 ROWS FETCH NEXT @__p_1 ROWS ONLY)
ORDER BY [p].[BlogsId]
```

***

##### If ` .Skip(1)` is changed to ` .Skip(0)`

### Result

```json
[{"blogName":"Foo","postTitles":["Lorem","Ipsum"]}]
```

### EF Generated SQL

No changes.

***

##### If `.Select(b => new { BlogName = b.Name, PostTitles = FindTitles(b) })`

##### is changed to `.Select(b => new { BlogName = b.Name, PostTitles = b.Posts.Select(p => p.Title) })`

<sub>(here ` .Include(b => b.Posts)` is useless)</sub>

### Result

```json
[{"blogName":"Bar","postTitles":["Dolor"]}]
```

### EF Generated SQL

```sql
SELECT [b].[Name], [b].[BlogsId]
FROM [Blogs] AS [b]
ORDER BY @@ROWCOUNT
OFFSET @__p_0 ROWS FETCH NEXT @__p_1 ROWS ONLY

SELECT [p].[BlogsId], [p].[Title]
FROM [Posts] AS [p]
```

***

Obviously what's happening is that the code directly inside the `.Select()` lambda is interpreted differently. Is there a way to keep this "context" inside another function (in this case, `FindTitles()`)?

**Comment by wc-matteo** [comment]
As a note, not related to this issue, calling a function inside `.Select()` (e.g. `FindTitles()`) while passing an entity as an argument (e.g. `Blogs)`, increases the response time from 80 ms to 300 ms, in my case; is sufficient to pass/not pass the argument to trigger the difference.

Is this something you're aware of? Am I doing it wrong?

**Comment by smitpatel** [comment]
EF generates following query in latest codebase
```
exec sp_executesql N'SELECT [p].[PostId], [p].[BlogId], [p].[Title]
FROM [Posts] AS [p]
INNER JOIN (
    SELECT DISTINCT [t].*
    FROM (
        SELECT [b].[BlogId]
        FROM [Blogs] AS [b]
        ORDER BY [b].[BlogId]
        OFFSET @__p_0 ROWS FETCH NEXT @__p_1 ROWS ONLY
    ) AS [t]
) AS [b0] ON [p].[BlogId] = [b0].[BlogId]
ORDER BY [b0].[BlogId]',N'@__p_0 int,@__p_1 int',@__p_0=1,@__p_1=1
```
Above query gives answers as expected. Fixed in #6002. 
As for passing function, the difference happens in execution that, when you use function, it will run include hence 2 queries (the 2nd one being inner join query). If you use inline projection then the projection is known to EF which does not require all data from post, and it uses n+1 queries (1 query for each blog id included.) so with `.Take(1)` it fires much smaller query hence faster. If you take larger number of entity then it would perform slow.

**Comment by wc-matteo** [comment]
Thank you for the answer @smitpatel.
Regarding the difference in perf when passing an entity as an argument, I have another piece of code where there are no `.Include()` and, again, just passing an entity as an argument to an empty function inside a `.Select()` have a major impact in response times. And the queries against the database remain practically
the same, with or without the argument.
Is this something you recommend against doing? Should I try to create a repro and file an issue?
